design/004-datafrom-key-rewrite.md

```diff
@@ -0,0 +1,103 @@
+
+## Motivation
+It is hard to administer key creation without having weird names that convolutes with the path. It is natural to have a feature to allow secret Keys manipulation both from the lifecycle of the secret management and as part of the templating feature.
+
+### Goals
+- CRD Design for dataFrom key rewrite
+
+### Non-Goals
+Do not implement secret values rewrite, nor handle with conversion.
+
+### Terminology
+- Secret Key: the kubernetes secret key, aka the `string` in `map[string][]byte` in the codebase used to render Kubernetes Secrets.
+
+## Proposal
+
+Add a rewrite step after applying either `GetSecretMap` or `GetAllSecrets`, but before applying any conversion logic.
+
+### User Stories
+1. As an user I want to be able to remove paths from my Secret Keys
+2. As an user I want to be able to customize secret keys to have more meaningful value to the tenants consuming it.
+3. As an user I want external-secrets to tell me if there is any rewrite conflict and not to overwrite secrets information.
+
+### API
+Proposed CRD changes:
+
+```yaml
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+metadata:
+  name: sample
+  namespace: default
+spec:
+  refreshInterval: 1m
+  target:
+   name: foobar
+  secretStoreRef:
+    name: secretstore-sample
+    kind: SecretStore
+  dataFrom:
+    - extract:
+        key: /foo/bar
+      rewrite:
+      -  origin: my-known-key
+         target: new_key_name
+      -  origin: my-unknown-(.*)
+         target: unknown-$1
+    - find:
+        conversionStrategy: Unicode
+        path: "my/path/"
+        name:
+          regexp: "my-pattern"
+      rewrite:
+      -  origin: my/path/(.*)
+         target: $1
+      -  origin: creds/(db|system)/(.*)
+         target: $2-$1
+    - find:
+        conversionStrategy: Default
+        path: "my/path/"
+        tags:
+          env: dev
+      rewrite:
+      -  origin: my/path/(.*)
+         target: dev-$1
+
+```
+
+### Behavior
+After applying `GetAllSecrets` or `GetSecretMap`, a rewrite logic is applied by using regular expressions capture groups. This will convert the Secret Keys according to a given standard, possibly even up to a well known key for the `GetSecretMap`. After the key rewrite is done, the `ConversionStrategy` should be applied, and then the new collection of Secret Keys (`map[string][]byte`) can be rendered to template.
+
+### Drawbacks
+
+It would not be trivial to replace existing characters for new characters with this strategy. This could be implemented with a `replace` method afterwards.
+
+### Acceptance Criteria
++ If multiple keys have the same name an error should happen to the user referencing the original Secret Keys.
```

Comment:
Does it mean that no merging will happen in case I want to layer env-specific params on top of some defaults? For example:

```
    - find:
         conversionStrategy: Unicode
         path: "/my-app/defaults"
         name:
           regexp: "*"
    - find:
         conversionStrategy: Unicode
         path: "/my-app/production"
         name:
           regexp: "*"
```

I would expect Production values to override any defaults. Does this proposal result in such behaviour, or will ESO simply fail to sync in case of identical keys in 2 different paths (for example in SSM)?

Reply:
hey @Fodoj good point! In it's current state the controller errors out if there is a key collision (there's no rewrite yet). Looking forward i think this is a valuable feature to have that allows users to compose their configuration parameters.

Reply:
I think we can make it that after a rewrite, if keys did match, to just set the new value instead of checking for a collision. I'm just not sure if we want to do this. 

For instance, if a single key stops existing in the find operation from the production side, it can be hard to troubleshoot.

Also, we need to take care to avoid inducing a breaking change of the current behaviour (the abort logic is already present on `utils.ConvertKeys` method).

Reply:
Agreed with that! This also allows us to move along with this one! :+1: